### PR TITLE
feat/rooms Join/Leave Room Endpoint

### PR DESCRIPTION
**Pull request by Anphs**
POST Join Room Endpoint
POST Leave Room Endpoint
Room Leader Mechanism

### CODE REVIEW

File: central-service/services/room.go [hunk 1]
```diff
@@ -54,24 +60,113 @@ func (service *RoomService) FindRoomByID(roomID string) (*models.Room, error) {
...
+				log.Printf("Room %s is empty\n", roomID)
+				if resultCmd := service.rdb.Del(context.Background(), key); resultCmd.Err() != nil {
+					log.Printf("Error deleting key %s: %v\n", key, resultCmd.Err())
+					return nil, resultCmd.Err()
+				}
+			} else {
+				return nil, err
+			}
+		}
+		room.Admin = result
```
Comment (DedsecKnight) [inline central-service/services/room.go]: I think you would need to save the room update here. Look into [this guide](https://gorm.io/docs/update.html) for more details.

File: central-service/services/room.go [hunk 2]
```diff
@@ -54,24 +60,113 @@ func (service *RoomService) FindRoomByID(roomID string) (*models.Room, error) {
...
+	}
+	return room, nil
```
Comment (DedsecKnight) [inline central-service/services/room.go]: Any reason why we need to return a Room object here?

File: central-service/services/room.go [hunk 3]
```diff
@@ -54,24 +60,113 @@ func (service *RoomService) FindRoomByID(roomID string) (*models.Room, error) {
...
+		return nil, err
+	}
+	if result < 1 {
+		return nil, RoomServiceError{Message: "Are you already in this room?"}
+	}
+	// if round is already started
+	// create a participant object
+	// notify RTC
+	log.Printf("Users in room %s:\n %v\n", roomID, service.rdb.ZRange(context.Background(), key, 0, -1))
+	return room, nil
```
Comment (DedsecKnight) [inline central-service/services/room.go:86]: Any reason why we need to return a Room object here?

File: central-service/services/room.go [hunk 4]
```diff
@@ -54,24 +60,113 @@ func (service *RoomService) FindRoomByID(roomID string) (*models.Room, error) {
...
 }
 
-type RoomNameError struct{
+// Join a room
+func (service *RoomService) JoinRoom(roomID string, userID string) (*models.Room, error) {
+	room, err := service.FindRoomByID(roomID)
+	if err != nil {
+		return nil, err
+	}
+	// add check to make sure they're not in other rooms?
```
Comment (DedsecKnight) [inline central-service/services/room.go]: Probably reasonable to assume that we allow a person to be in multiple rooms at the same time

**Review by DedsecKnight — CHANGES_REQUESTED** [review]

File: central-service/services/room.go (hunk 3 above)
Comment (Anphs) [reply central-service/services/room.go:86]: The Room object contains the room name if we want to display it in the UI along with the current admin id if we want to color their chat messages

File: central-service/services/room.go (hunk 2 above)
Comment (Anphs) [reply central-service/services/room.go]: Changed it to return error or nil

**Review by 1solomonwakhungu — APPROVED** [review]
Looks good.

Will need to setup a meeting to discuss communication between RTC and Central Service.

**Review by DedsecKnight — APPROVED** [review]
lgtm!

**Pull request merged**